packages/protocol/contracts/bridge/NFTVault.sol

```diff
@@ -0,0 +1,326 @@
+// SPDX-License-Identifier: MIT
+//  _____     _ _         _         _
+// |_   _|_ _(_) |_____  | |   __ _| |__ ___
+//   | |/ _` | | / / _ \ | |__/ _` | '_ (_-<
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+
+pragma solidity ^0.8.20;
+
+import {
+    IERC721Receiver
+} from "@openzeppelin/contracts/token/ERC721/IERC721Receiver.sol";
+import {
+    IERC1155Receiver
+} from "@openzeppelin/contracts/token/ERC1155/IERC1155Receiver.sol";
+import {
+    IERC165
+} from "@openzeppelin/contracts/utils/introspection/IERC165.sol";
+import {AddressResolver} from "../common/AddressResolver.sol";
+import {EssentialContract} from "../common/EssentialContract.sol";
+import {IBridge} from "./IBridge.sol";
+import {LibERC721} from "./erc721/libs/LibERC721.sol";
+import {LibERC1155} from "./erc1155/libs/LibERC1155.sol";
+
+/**
+ * This vault holds all ERC721 and ERC1155 tokens that users have deposited.
+ * It also manages the mapping between canonical ERC721/1155 tokens and their bridged
+ * tokens.
+ */
+contract NFTVault is 
+    EssentialContract,
+    IERC721Receiver,
+    IERC1155Receiver
+{
+    /* ***************
+     * Constants     *
+     *************** */
+    bytes4 constant ERC721_INTERFACE_ID = 0x80ac58cd;
+    bytes4 constant ERC721_METADATA_INTERFACE_ID = 0x5b5e139f;
+    bytes4 constant ERC721_ENUMERABLE_INTERFACE_ID = 0x780e9d63;
+
+    bytes4 constant ERC1155_INTERFACE_ID = 0xd9b67a26;
+    bytes4 constant ERC1155_METADATA_INTERFACE_ID = 0x0e89341c;
+
+    enum NFTType{ ERC721, ERC1155}
+
+    /*********************
+     * Structs           *
+     *********************/
+
+     struct CanonicalNFT {
+        uint256 srcChainId;
+        address tokenAddr;
+        string symbol;
+        string name;
+        string uri;
+        NFTType nftType;
+     }
+
+    struct SendNFTOpts {
```

Comment:
I think for ERC20, ERC712, and ERC1155, we should define different structs but we can name the all the same `BridgeTransferOp` (change ERC20Vault's `MessageDeposit` to it as well).